compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -85,6 +86,64 @@ upgradeTemplates n =
     , "    convert B." <> n <> "{..} = A." <> n <> " {..}"
     ]
 
+-- | Generate the source for a package containing template instances for all templates defined in a
+-- package.
+generateInstancesPkgFromLf ::
+       (LF.PackageRef -> UnitId)
+    -> LF.PackageId
+    -> LF.Package
+    -> [(NormalizedFilePath, String)]
+generateInstancesPkgFromLf getUnitId pkgId pkg =
+    catMaybes
+        [ generateInstanceModule
+            Env
+                { envGetUnitId = getUnitId
+                , envQualify = False
+                , envMod = mod
+                }
+            pkgId
+        | mod <- NM.toList $ LF.packageModules pkg
+        ]
+
+-- | Generate a module containing template/generic instances for all the contained templates.
+generateInstanceModule ::
+       Env -> LF.PackageId -> Maybe (NormalizedFilePath, String)
```

Comment:
This whole function is almost identical to `generateGenInstanceModule` apart from the actual generation of the instances which is a separate function anyway. Can we factor out the common code?

Reply:
They are also slightly different in the qualification. I find harder to read and understand when I factor it out.